homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +148,192 @@
-            _setup_device(hass, config, ipaddr, device_config)
+    return unload_ok
 
-    return True
+
+async def _async_setup_device(hass: HomeAssistant, ipaddr: str, config: dict,) -> None:
+    # Set up device
+    bulb = Bulb(ipaddr, model=config.get(CONF_MODEL) or None)
+    capabilities = await hass.async_add_executor_job(bulb.get_capabilities)
+    if capabilities is None:  # timeout
+        _LOGGER.error("Failed to get capabilities from %s", ipaddr)
+        raise ConfigEntryNotReady
+    device = YeelightDevice(hass, ipaddr, config, bulb)
+    await hass.async_add_executor_job(device.update)
+    await device.async_setup()
+    return device
 
 
-def _setup_device(hass, _, ipaddr, device_config):
-    devices = hass.data[DATA_YEELIGHT]
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
 
-    if ipaddr in devices:
-        return
 
-    device = YeelightDevice(hass, ipaddr, device_config)
+class YeelightScanner:
+    """Scan for Yeelight devices."""
 
-    devices[ipaddr] = device
-    hass.add_job(device.setup)
+    _scanner = None
+    _get_lock = asyncio.Lock()
+
+    @staticmethod
+    async def async_get(hass: HomeAssistant):
+        """Get scanner instance."""
+        with await YeelightScanner._get_lock:
+            if YeelightScanner._scanner is None:
+                YeelightScanner._scanner = YeelightScanner(hass)
+        return YeelightScanner._scanner
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize class."""
+        self._hass = hass
+        self._seen = {}
+        self._seen_lock = asyncio.Lock()
+        self._callbacks = {}
+        self._scanning = False
+
+    async def _async_scan(self):
+        _LOGGER.debug("Yeelight scanning")
+        # Run 3 times as packets can get lost
+        for _ in range(3):
+            devices = await self._hass.async_add_executor_job(discover_bulbs)
+            for device in devices:
+                with await self._seen_lock:
+                    unique_id = device["capabilities"]["id"]
+                    if unique_id in self._seen:
+                        continue
+                    ipaddr = device["ip"]
+                    self._seen[unique_id] = ipaddr
+                    _LOGGER.debug("Yeelight discovered at %s", ipaddr)
+                    if unique_id in self._callbacks:
+                        self._hass.async_create_task(self._callbacks[unique_id](ipaddr))
+                        self._callbacks.pop(unique_id)
+                        if len(self._callbacks) == 0:
+                            self._async_stop_scan()
+
+        if not self._scanning:
+            return
+
+        await asyncio.sleep(SCAN_INTERVAL.seconds)
+        if self._scanning:  # make sure we are still scanning after sleep
+            self._hass.async_create_task(self._async_scan())
```

Comment:
We need to save the scan task so we can cancel it when needed without waiting for the scan interval.